bonsai-core/src/main/scala/com/stripe/bonsai/IndexedBitSet.scala

```diff
@@ -160,20 +244,53 @@ object IndexedBitSet {
     }
 
   /**
-   * TODO: maybe tom can explain this?
+   * Returns the rank of the i-th bit in a 32-bit word.
    */
-  def rankWord(word: Int, i: Int): Int = {
+  private[bonsai] def rankWord(word: Int, i: Int): Int = {
     val mask = ~(-1L << ((i + 1))).toInt
     java.lang.Integer.bitCount(word & mask)
   }
 
   /**
+   * Returns the index of the i-th set bit in a 32-bit word.
+   *
+   * @param word a 32-bit word to search
+   * @param rank a value in [1,32] (inclusive)
+   */
+  private[bonsai] def selectWord(word: Int, rank: Int): Int = {
```

Comment:
One idea would be to do a bitCount manually, but store the intermediate steps. From there, we have the total bit count, the bit count in each half, the bit count in each quarter, etc. So, then the binary search just uses some simple masks/shifts, rather than recomputing the bit count each time. So, one bit count is enough to do the binary search. I may try that...